csrc/device_lower/utils.cpp

```diff
@@ -1013,6 +1016,618 @@ bool predicateAtEnd(ForLoop* loop) {
+    return related(g, to);
+  });
+}
+
+bool related(const std::monostate& current, const ValGroup& to) {
+  return false;
+}
+
+bool related(const LinearGroupProjection& current, const ValGroup& to) {
+  return LinearGroupProjection::dispatch(
+      [&](const auto& current) { return related(current, to); }, current);
+}
+
+// Given an expression on the traversal path and its direction, get the from and
+// to groups. Note that the traversal path is obtained by running BFS from
+// domain to linear_g, so the direction is flipped with respect to how we
+// propagate from linear_g to domain.
+auto fromGroups(
+    const ValGraph& id_graph,
+    const ExprGroup& eg,
+    Direction direction) {
+  return direction == Direction::Forward ? id_graph.outputGroups(eg)
+                                         : id_graph.inputGroups(eg);
+}
+
+auto toGroups(
+    const ValGraph& id_graph,
+    const ExprGroup& eg,
+    Direction direction) {
+  return direction == Direction::Forward ? id_graph.inputGroups(eg)
+                                         : id_graph.outputGroups(eg);
+}
+
+// Do the propagation from linear_g to domain.
+LinearGroupProjection propagate(
+    const LinearGroupProjection& current,
+    const ValGraph& id_graph,
+    const ExprGroup& eg,
+    Direction direction);
+
+LinearGroupProjection propagate(
+    const ValGroup& current,
+    const ValGraph& id_graph,
+    const ExprGroup& eg,
+    Direction direction) {
+  auto from = fromGroups(id_graph, eg, direction);
+  auto to = toGroups(id_graph, eg, direction);
+  if (from.size() == 1) {
+    // If linear_g is equivalent to from, and from is split into two groups,
+    // then these two groups together represents linear_g.
+    NVF_ERROR(to.size() == 2);
+    NVF_ERROR(from.front() == current);
+    return Composition<LinearGroupProjection>{to.front(), to.back()};
+  } else {
+    // If linear_g is merged with another group, then part of the merged group
+    // represents linear_g.
+    NVF_ERROR(from.size() == 2);
+    NVF_ERROR(to.size() == 1);
+    NVF_ERROR(from.front() == current || from.back() == current);
+    return PartOf<LinearGroupProjection>{
+        std::make_shared<LinearGroupProjection>(to.front()),
+        /*inner_extent=*/from.front() == current
+            ? from.back()->front()->as<IterDomain>()->extent()
+            : nullptr,
+        /*selected_extent=*/
+        simplifyExpr(current->front()->as<IterDomain>()->extent())};
+  }
+}
+
+LinearGroupProjection propagate(
+    const PartOf<LinearGroupProjection>& current,
+    const ValGraph& id_graph,
+    const ExprGroup& eg,
+    Direction direction) {
+  auto from = fromGroups(id_graph, eg, direction);
+  auto to = toGroups(id_graph, eg, direction);
+  if (from.size() == 1) {
+    NVF_ERROR(to.size() == 2);
+    NVF_ERROR(related(*current.group, from.front()));
+    auto group = propagate(*current.group, id_graph, eg, direction);
+    if (!group.hasValue()) {
```

Comment:
When does this happen? IIUC, we are propagating the group of `current`, which is not `linear_g`.

Reply:
Added a comment in code:
```C++
      // Propagation of group may fail. For example, if the group before
      // propagation is [4, 3], and eg is split(3, 2), then the end result
      // would be like [4, 2, 2]. However, because the split is indivisible,
      // this makes group discontiguous. That is, "a composition of 4 and 3"
      // is not a "composition of 4, 2, and 2". In this case, the return value
      // of `propagate` will be std::monostate, indicating that the linear
      // proving fails. For this case, we just propagate the failure.
```